### PR TITLE
Refactor prover api

### LINKED ISSUE #92: Refactor prover to use the risc0-zkvm API
Currently the Foundry template includes code to call Bonsai with the Bonsai SDK. Proving with Bonsai is also provided under the unified API with `risc0-zkvm`, and this API is a much better user experience. We should refactor the template to use the `risc0-zkvm` API and drop the `BonsaiProver` type. This may require changes to the `risc0` monorepo.

Note that using the `risc0-zkvm` API does not immediately mean support for local proving. The blocker for local proving as an easily supported mode is the fact that the only prover option that works today is rapidsnark, and rapidsnark does not integrate nicely with our Rust workflows. As a result, we cannot yet provide a seemless local proving experience for Groth16, as some of the steps are manual and require Docker on x86 to run.

Resolved by: https://github.com/risc0/risc0-foundry-template/pull/106

### PR DESCRIPTION
**Pull request by capossele**
Closes #92

### CODE REVIEW

File: .github/workflows/main.yml
Comment (nategraf) [inline .github/workflows/main.yml:1]: Add `dev` to `branches` on triggers

**Comment by nategraf** [comment]
In order to test this against `risc0` main, I did some work to update a few things https://github.com/risc0/risc0-foundry-template/pull/109

**Comment by nategraf** [comment]
Note: I also force-pushed the `dev` branch to match the commit history of `main`. I think ideally, we'll have `main` lag behind `dev`, but share `git` history with it.

File: .gitmodules [hunk 1]
```diff
@@ -13,4 +13,4 @@
 [submodule "lib/risc0-ethereum"]
 	path = lib/risc0-ethereum
 	url = https://github.com/risc0/risc0-ethereum
-	branch = release-0.9
+	branch = capossele/refactor-zkvm-usage
```
[the remark this answers is not in the document]
Comment (capossele) [reply .gitmodules:16]: Yeah sure now that all the changes have been merged on risc0-ethereum it’s not required anymore but before that it was using the branch with the required changes to work (e.g., the updated ffi mainly)

[the remark this answers is not in the document]
Comment (nategraf) [reply .gitmodules:16]: So what I'm actually trying to point out is that I think this config line has no effect. What actually seems to matter is the commit recorded in the sub-module reference.

[the remark this answers is not in the document]
Comment (capossele) [reply .gitmodules:16]: I see, that’s the way I use to do that. Is there a better way to specify a given sub-module reference? Or what you’re suggesting is to use the same way but avoid committing the config change after the submodule reference has been updated?

[the remark this answers is not in the document]
Comment (nategraf) [reply .gitmodules:10]: Ok, so it appears this is used by `forge update`. Let's keep this and make sure to update it to keep it in sync with the checked out commit in the sub-module.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -15,12 +15,16 @@ concurrency:
 env:
   GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
   RUST_BACKTRACE: "1"
-  RISC0_CRATE_VERSION: "^0.21"
-  RISC0_TOOLCHAIN_VERSION: v2024-02-08.1
+  RISC0_MONOREPO_REF: "main"
+  RISC0_TOOLCHAIN_VERSION: v2024-04-22.0
```
[the remark this answers is not in the document]
Comment (nategraf) [reply .github/workflows/main.yml]: I made this change in https://github.com/risc0/risc0-foundry-template/pull/106/commits/c2802300b1e5880eaf60a442f28d13c978de0fda

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -15,12 +15,16 @@ concurrency:
...
 
 jobs:
   test:
     runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        feature: [default]
```
[the remark this answers is not in the document]
Comment (nategraf) [reply .github/workflows/main.yml]: I made this change in https://github.com/risc0/risc0-foundry-template/pull/106/commits/c2802300b1e5880eaf60a442f28d13c978de0fda

File: .github/actions/cargo-risczero-install/action.yaml
[the remark this answers is not in the document]
Comment (nategraf) [reply .github/actions/cargo-risczero-install/action.yaml:1]: I've updated this in https://github.com/risc0/risc0-foundry-template/pull/106/commits/c2802300b1e5880eaf60a442f28d13c978de0fda

**Comment by nategraf** [comment]
@capossele, I pushed a few commits to this branch. Definitely let me know if you disagree with any of the changes.

**Comment by nategraf** [comment]
In the latest commits, I am using the `cargo-risczero-install` action from https://github.com/risc0/risc0-ethereum/pull/128

**Review by nategraf — APPROVED** [review]

**Pull request merged**